### ISSUE TITLE
Persistent Indexes?

### ISSUE THREAD

**Issue by nevf** [Ivshti/linvodb3#35]
Are there any plans for persistent indexes stored in leveldb?

**Comment by Ivshti** [comment]
Yep, there are plans for those. However, they cannot be automatic like the in-memory indexes, and to be truly useful, compound indexes must be implemented.

But it's a great feature, since it could allow larger datasets and remote access/replication without waste of resource

**Comment by nevf** [comment]
@Ivshti Thanks for the prompt reply. What do you mean by "cannot be automatic ..."? And yes compound indexes are a must.  Do you have a time-frame for this?

An in-process Node database package which was compatible with Mongodb would be incredibly useful, to me and I'm sure many others. I'm currently using Mongodb in my Web Knowledge Base app Clibu http://clibu.com and it is great. However I'm also packaging up a stand-alone version of Clibu which can be installed on the end-users PC. This is where an in-process db would be so much better than using Mongodb. 

Thanks for your work on linvodb3.

**Comment by Ivshti** [comment]
"cannot be automatic" - currently indexes are created automatically for every looked-up property; if we do this for persistent indexes, the user may accidently polute everything with indexes 

Also, currently compound indexes aren't given a priority as with <1 million datasets and in-RAM indexes, there's no real benefit from compound indexes. But I agree, they are important, especially with persistent indexes.

Regarding MongoDB compatibility - unfortunately, the goal here is to be similar to mongodb, not 100% compatible. 100% compatible is too responsible, difficult and would result in some technical compromise. Usually this shouldn't be a problem for most cases, but just keep that in mind.

I don't have much time lately for this project, but I think in 2-3 months I will absolutely have to implement persistent indexes

**Comment by nevf** [comment]
@Ivshti Thanks for the extra info. I didn't realize linvodb3 created an index for each queried property. You definitely wouldn't want that for persistent indexes. If I understand correctly I'd use ```{autoindex: false}``` to turn that off and use ```Doc.ensureIndex()``` to explicitly create just the indexes I want.

I understand the MongoDB compatibility and that's fine.

You maybe interested in reading http://pouchdb.com/2014/05/01/secondary-indexes-have-landed-in-pouchdb.html - "For the database geeks: implementation details" which discusses how they've added support for secondary indexes to PouchDB. In essence that concatenate the secondary key + id of the document being indexed, which means they can update these secondary keys efficiently.

There are also some levelup libs such as https://github.com/juliangruber/level-secondary and https://github.com/eugeneware/jsonquery-engine which implement secondary indexes which could be worth looking at.

I'll look forward to seeing you future release with persistent + secondary + compound indexes.

**Comment by Ivshti** [comment]
@nevf yes - if you autoindex: false, no indexes are going to be created. And you can use ensureIndex then

Thanks for the hints, I'll check them out!

**Comment by wgerven** [comment]
Hello @Ivshti, very interesting database and thanks for the sharing your work. I am considering it for integration in an Electron desktop app. 
I am trying to get a better understanding of your database from the documentation. Could you please comment on what is written here https://docs.omniref.com/js/npm/linvodb3/3.5.0#label-Indexing 

> If your datastore is persistent, the indexes you created are persisted in the datafile, when you load the database a second time they are automatically created for you. No need to remove any ensureIndex though, if it is called on a database that already has the index, nothing happens.

and the nature of this conversation you had with @nevf? Do I understand that in the mean time since January you implemented persisting of secondary indexes on disk? In the quoted text, what does _'they are automatically created'_ exactly mean? 

Thank you.
